### PR TITLE
Label Ensembl canonical on Transcript Image using data from Thoas

### PR DESCRIPTION
**Pull request by jyothishnt**
## Type

- [ ] Bug fix
- [ ] New feature
- [x] Improvement to existing feature

## Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-520

## Description
The default (topmost) transcript in the transcripts table will be identified with a label:

1) Labelled transcript as 'MANE Select' or 'Ensembl canonical' depending on the evidence used to identify it
2) Added logic to display additional 'MANE Plus Clinical' transcripts if any
3) Tooltip updated from the definition field coming from metadata

~NOTE: Currently it is not the topmost, we will have to tweak the sorting algorithm which will be done in another ticket.~

## Deployment URL
MANE select - http://add-ensembl-canonical.review.ensembl.org/entity-viewer/homo_sapiens_GCA_000001405_28/gene:ENSG00000139618?view=transcripts

Ensembl canonical (not Mane select) http://add-ensembl-canonical.review.ensembl.org/entity-viewer/homo_sapiens_GCA_000001405_28/gene:ENSG00000278704?view=transcripts

Multiple MANE labels http://add-ensembl-canonical.review.ensembl.org/entity-viewer/homo_sapiens_GCA_000001405_28/gene:ENSG00000187098?view=transcripts

## Views affected
Entity Viewer

### Other effects
NA

- [ ] I have created any required tests
- [ ] I have checked any requirements for Google Analytics
- [ ] The PR has as its final commit a WASM/JS update commit if any Rust files were updated.

## Possible complications
NA

### CODE REVIEW

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/DefaultTranscriptsList.tsx
Comment (ens-ma7) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/DefaultTranscriptsList.tsx:83]: i think this code need to stay because we need to apply the default sorting first and then display the transcripts. sorry if i confused you when we spoke but we do need this default sorting and when you apply the default sorting, the first transcript will be either "Mane" or "Canonical" or can be empty.

File: src/ensembl/src/shared/types/thoas/metadata.ts
Comment (ens-ma7) [inline src/ensembl/src/shared/types/thoas/metadata.ts:17]: according to thoas docs, there is also a description field. Do you not need it? As discussed i have created my branch for the transcript more info from this branch and i am going to add it to the code unless its not needed.

**Comment by azangru** [comment]
For the reference, the alternative way of writing the default sorter function we tried was:

```
export function defaultSort<
  T extends Array<
    IsProteinCodingTranscriptParam & {
      slice: SliceWithOnlyLength;
      so_term: string;
    }
    & Pick2 <FullTranscript, 'metadata', 'canonical' | 'mane'>
  >
>(transcripts: T): T {

  const sortingFunction = (a: T[0], b: T[0]): number => {
    if (b.metadata.canonical) {
      // canonical transcript goes first
      return 1;
    } else if (a.metadata.canonical) {
      return -1;

    } else if (b.metadata.mane && !a.metadata.mane) {
      // transcripts with MANE metadata go second
      return 1;
    } else if (a.metadata.mane) {
      return -1;


    } else if (isProteinCodingTranscript(b) && !isProteinCodingTranscript(a)) {
      // protein-coding transcripts go before non-coding
      return 1;

    } else if (isProteinCodingTranscript(a) && !isProteinCodingTranscript(b)) {
      // protein-coding transcripts go before non-coding
      return -1;

    } else {
      // transcripts within a group are ranked by their length (from longer to shorter)
      return compareTranscriptLengths(a, b);
    }
  };

  return [...transcripts].sort(sortingFunction) as T;
}
```

I was expecting it to be several times faster than the other sorting function (because it was going through the array just once), but it ended up being slower. Still not sure how that happened. But I am just recording the alternative implementation here if we ever wanted to return to it.

File: src/ensembl/src/shared/types/thoas/metadata.ts
Comment (azangru) [reply src/ensembl/src/shared/types/thoas/metadata.ts:17]: > Do you not need it?

It isn't populated with any real data at the moment, and we aren't using it in the client anyway. Currently, the definition field is sufficient for us.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/DefaultTranscriptsList.tsx
Comment (ens-ma7) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/DefaultTranscriptsList.tsx:83]: Agreed we dont need this piece of code but will be good to use `useMemo` to apply sorting when the geneId or sortingRule are different

File: src/ensembl/src/content/app/entity-viewer/gene-view/GeneView.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/GeneView.tsx:160]: We don't really use this value anywhere 😃

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/default-transcripts-list-item/DefaultTranscriptListItem.test.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/default-transcripts-list/default-transcripts-list-item/DefaultTranscriptListItem.test.tsx:91]: The test isn't really checking what it claims to be checking. The description of the test talks about Ensembl canonical transcript, but the code on line 95 is checking the mane metadata (which means that this also happens to be a MANE select transcript)?

If you really want to test the content of the label, you should probably test the `TranscriptQualityLabel` component, move this test over there, and add some other test cases:

- canonical + mane select => uses transcript.metadata.mane.label
- canonical without mane select => uses transcript.metadata.canonical.label
- non-canonical, with mane metadata => uses transcript.metadata.mane.label
- non-canonical, no mane metadata => returns null

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx:109]: Where do you use all this metadata for gene external references? You probably only need canonical.value and mane.value for sorting?

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/tests/transcripts-sorter.test.ts
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/shared/helpers/tests/transcripts-sorter.test.ts:29]: Umm, I am not sure I like this function, for several reasons, one of which being that it claims to create a transcript with empty metadata, but only empties two metadata fields.

Why don't we assign nulls to all metadata fields directly in the `createTranscript` function?

File: src/ensembl/src/content/app/entity-viewer/shared/components/default-transcript-label/TranscriptQualityLabel.tsx
Comment (imransl) [inline src/ensembl/src/content/app/entity-viewer/shared/components/default-transcript-label/TranscriptQualityLabel.tsx:20]: We need to separate this from `QuestionButton` import which is a component

File: src/ensembl/src/content/app/entity-viewer/shared/components/default-transcript-label/TranscriptQualityLabel.tsx
Comment (imransl) [inline src/ensembl/src/content/app/entity-viewer/shared/components/default-transcript-label/TranscriptQualityLabel.tsx:45]: How about destructuring the `metadata` properties so that we don't need to repeat `props.metadata` everywhere?

```const { canonical, metadata, mane } = props.metadata;```

**Review by ens-ma7 — APPROVED** [review]

**Comment by ens-ma7** [comment]
@jyothishnt i have approved from what i can understand, please review the suggested changes by @azangru and @imransl and wait for their approval before merging. Dont rely on my approval :)

File: src/ensembl/src/content/app/entity-viewer/shared/helpers/tests/transcripts-sorter.test.ts
Comment (jyothishnt) [reply src/ensembl/src/content/app/entity-viewer/shared/helpers/tests/transcripts-sorter.test.ts:29]: updated code, please check.

File: src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx
Comment (jyothishnt) [reply src/ensembl/src/content/app/entity-viewer/gene-view/components/gene-view-sidebar/external-references/GeneExternalReferences.tsx:109]: yeah you are right. It is used for default sort in this component. updated

**Review by azangru — APPROVED** [review]

**Pull request merged**
